### DISCUSSION TITLE
How to get attribute value from a group?

### DISCUSSION THREAD

**Discussion by haziq-zur** [LiorBanai/HDF5-CSharp#279] [Q&A] [answered]
Hi all,

I have a .h5 file generated from an instrument and doing a hdf5 file parser to extract the content and expose it into a map. The file structure that I have is as below.
![image](https://user-images.githubusercontent.com/52647283/222021329-a289c9b6-ebc0-48f8-a8c7-070bd640a74a.png)

 I manage to get the array of data from the dataset, but I try to get the attribute of the parent I'm having an exception -> 
 System.ArgumentException: 'Must provide at least one rank.'

I use the method `hdf5.ReadAttribute<int>(wvf_group_id, "NumWaveforms")`

**Comment by flappah** [comment]
No answer but just to add at this, it also happens to me. I try to read an attribute of an enumerated type but am only interested in the byte value of the attribute and not the enumeration. So I use the same method as OP. And then the same error occurs. "Must provide at least one rank".

**Reply by LiorBanai** [reply]
Hi @haziq-zur and @flappah can you attach an example file for me to test? I'll try to get to it in few days

**Reply by flappah** [reply]
Hi Lior, 

sure! This is the code I use. It is more or less equivalent to the other code I posted only then without going to the underlying HDF group p/invoke wrapper:

```
            var fileId = Hdf5.OpenFile(fileName);
            var groupId = Hdf5.CreateOrOpenGroup(fileId, "/SurfaceCurrent");
            byte dataCodingFormat = 0;
            try
            {
                dataCodingFormat = Hdf5.ReadAttribute<byte>(groupId, "dataCodingFormat");
            }
            finally
            {
                Hdf5.CloseGroup(groupId);
                Hdf5.CloseFile(fileId);
            }
```

**Reply by flappah** [reply]
BTW, it's part of a project I'm currently doing that can read IHO's S1xx standards. Right now I'm implementing S-111 that needs to read data from HDF5. To get a feel for the broader context it's on https://github.com/flappah/S1XViewer in the file https://github.com/flappah/S1XViewer/blob/master/S1XViewer.HDF/S111ProductSupport.cs

**Reply by haziq-zur** [reply]
Hi @LiorBanai,

I have attached the file that I used to extract data below. 
[cwaveformfloat.zip](https://github.com/LiorBanai/HDF5-CSharp/files/10904188/cwaveformfloat.zip)

Using the HDFView Tools, I can clearly see that there are some attribute in the /Waveforms group as posted previously. I did try using `Hdf5.ReadTreeFileStructure()` and `Hdf5.ReadFlatFileStructure()` to get the whole structure of the hdf5 file but the attribute content is showing zero which is weird since the HDFView can clearly show the attribute and its value.
![image](https://user-images.githubusercontent.com/52647283/223293261-6a8e5dc9-b3af-44a4-9ddc-11100ab140fc.png)

**Comment by flappah** [comment, marked as answer]
BTW, I managed to bypass the issue by just going to the underlying HDF-Group C wrapper. In my specific example I needed to retrieve the byte value (unsigned int 8 bit) which is the actual storage of the enumerated attribute type. The following code does this without a problem:

```
        public override int GetDataCodingFormat(string fileName)
        {
            var fileId = H5F.open(fileName, H5F.ACC_RDONLY, H5P.DEFAULT);
            var groupId = H5G.open(fileId, Encoding.ASCII.GetBytes("/SurfaceCurrent"), H5P.DEFAULT);
            var attribId = H5A.open(groupId, Encoding.ASCII.GetBytes("dataCodingFormat"), H5P.DEFAULT);
 
            var value = new byte[2];
            var handle = GCHandle.Alloc(value, GCHandleType.Pinned);
            try
            {
                if (H5A.read(attribId, H5T.NATIVE_UINT8, handle.AddrOfPinnedObject()) == 0)
                {
                    var returnValue = BitConverter.ToInt16(value, 0);
                    return returnValue;
                }
            }
            finally
            {
                handle.Free();
                H5A.close(attribId);
                H5G.close(groupId);
                H5F.close(fileId);
            }
 
            return -1;
        }

```
Maybe this can also help you out! 

[edited: made code sample complete]

**Reply by haziq-zur** [reply]
Hi @flappah,

I have tried your suggestion on using the underlying C wrapper to get the attribute and manage to get the attribute value that I want. For now, it is a matter of creating a looping function to iterate over all the attributes present in the groups/datasets. Thanks a lot for your help. For now, I will mark this as answer as I can use it to achieve my goals. But would like to know more on what happen with HDF5-CSharp not able to directly get the attribute. Thanks again.
